### PR TITLE
Pika transport class

### PR DESCRIPTION
**Pull request by ndevenish**
This is a rather large WIP branch to make a working pika (AMQP) transport. It builds off the also-WIP state of Markus' [pika_rewrite](https://github.com/DiamondLightSource/python-workflows/tree/pika-rewrite) branch. Done here:

- `_PikaThread` is rewritten to use pika's [`BlockingConnection`](https://pika.readthedocs.io/en/stable/modules/adapters/blocking.html), effectively making the IO loop owned by us instead of coordinating multiple threads. This part is possibly optional.
- `_PikaThread` manages reconnection, if possible, reconnecting queues and subscriptions if connections are dropped while receiving
- All the public-facing methods of `_PikaThread` are thread-safe. Operations which can be validated once completed return [Future](https://docs.python.org/3/library/concurrent.futures.html#future-objects) objects (the transport class appears to be synchronous for many operations, which means that these need to be waited on)
- The `test_pika.py` tests are rewritten and have been added to, so that:
  - `PikaTransport` tests are mocking out `_PikaThread` behaviour instead of directly mocking pika
  -  There are `test_pikathread_*` tests that test `_PikaThread`, using a live RabbitMQ dev server
- The `PikaTransport` class has been reworked using `_PikaThread` to manage the connection lifecycle and do the pika-talking parts
  
Non-exhaustive list of things still to be done/improved/tested:
- [ ] Either mock testing for `_PikaThread` when there is no dev-server, or better configuration for this
- [ ] Transaction support. This possibly needs knowledge of how RabbitMQ's [transaction semantics](https://www.rabbitmq.com/semantics.html) differ from ActiveMQ's, and what the workflows package expects.
- [ ] Exhaustive lifecycle testing and handling
  - [ ] Connections dropped while sending
  - [ ] Connections dropped between message receive and acknowledgement
  - [ ] Failure states of server disappearing not fully understood
- [ ] Things that we have not decided how to handle/handled in the server yet
  - [ ] Message Delays. This is explicit in the Transport interface, but not part of base RabbitMQ. This will probably be achieved through [plugin](https://github.com/rabbitmq/rabbitmq-delayed-message-exchange) or chained [DLQ](https://web.archive.org/web/20150311233606/http://blog.james-carr.org/2012/03/30/rabbitmq-sending-a-message-to-be-consumed-later/) fallbacks, in which case the transport class might be responsible for configuring the queues (or at least, knowing where to redirect the messages instead)
  - [ ] (potential) reconnection of broadcast queues without losing messages
  - [ ] Message timestamps (this appears to be free in ActiveMQ, but possibly needs a community [plugin](https://github.com/rabbitmq/rabbitmq-message-timestamp) on RabbitMQ?) - if free on ActiveMQ, do we want to just fill this on otherwise sending messages?

### CODE REVIEW

**Comment by ndevenish** [comment]
The `Basic.Properties` sent to AMQP is now filled out with node timestamp upon sending.

I've added the ability to pass through non-mandatory flag to `_broadcast`, and made `.broadcast_status` set `mandatory=False`. This might need a general change for broadcasting - unlike ActiveMQ, the destination (exchange here) can have no bound queues, which fails with `mandatory=True`.

**Comment by Anthchirp** [comment]
Things to discuss later:
*  [_Future instances (..) should not be created directly except for testing._](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future) Is there a benefit in using `Future()` over [`Event()`](https://docs.python.org/3/library/threading.html#threading.Event)s? Presumably yes, because you can (and do) carry results and exceptions. Question the becomes: why is the quoted sentence there and is there a more appropriate object?
* The restart avoidance logic in `_PikaThread` seems unnecessary? `Thread()` objects can already only be started once. Presumably mixing the `start()` function with the `Thread.start` function made this necessary - I don't think these should or need to be mixed.
* `_PikaSubscriptionKind` is expressing more queue vs exchange rather than direct vs fanout
* Code doesn't work on Python 3.8 (subscripted ABCMeta object; missing future annotations import). Can we please keep this 3.7+ compatible at a minimum for now.
* Running a SampleConsumer doesn't actually work.

**Comment by Anthchirp** [comment]
On the topic of timestamps. I found only 3 places where message timestamps are used:
* `queue_drain` for display purposes only
* `dlq_purge` to generate directory and file names
* status monitor

Notably, the controller ignores message timestamps and uses the time of message processing instead.
That same logic can be used in the status monitor. The `queue_drain` use case is just eye candy, and can be removed.

The use in `dlq_purge` is more interesting. The timestamp here could be helpful to track down related log messages and tells you how long a message has been in the DLQ - except that it doesn't. It always refers to the first send operation, which may be close to when it has been NACKd, or it may not be. When RabbitMQ DLQs a message with the current configuration a bunch of headers are added:
```
x-death: count:1
              exchange:
              queue:ispyb_pia
              reason:rejected
              routing-keys:ispyb_pia
              time:1633592240
x-delivery-count:0
x-first-death-exchange:
x-first-death-queue:ispyb_pia
x-first-death-reason:rejected
```
That timestamp in there is of when the message is DLQd, which is a much better indicator on when to focus in logs.

So in summary: On RabbitMQ we don't actually need timestamps. Therefore I think we should remove them.

File: src/workflows/transport/pika_transport.py [hunk 1]
```diff
@@ -760,7 +760,8 @@ def stop(self):
         # We might be waiting for an event, so give the event loop one...
         # We might already be closed or shutting down, so ignore errors for that
         try:
-            self._connection.add_callback_threadsafe(lambda: None)
+            if self._connection:
+                self._connection.add_callback_threadsafe(lambda: None)
```
[the remark this answers is not in the document]
Comment (Anthchirp) [reply src/workflows/transport/pika_transport.py:764]: :thinking:
